godot-macros/src/class/godot_api.rs

```diff
@@ -230,10 +233,18 @@ fn process_godot_fns(decl: &mut Impl) -> Result<(Vec<FuncDefinition>, Vec<Functi
             }
 
             match attr.ty {
-                BoundAttrType::Func { rename } => {
+                BoundAttrType::Func { rename, gd_self } => {
                     // Signatures are the same thing without body
-                    let sig = util::reduce_to_signature(method);
-                    func_definitions.push(FuncDefinition { func: sig, rename });
+                    let mut sig = util::reduce_to_signature(method);
+                    if gd_self {
+                        sig.params = sig.params.clone();
+                        sig.params.inner.remove(0);
```

Comment:
Before just removing the first element, we should:
* Ensure that there _is_ such an element at all (this panics otherwise).
* Ensure that the element _is_ of type `Gd<Self>`.
   * The complex way: I wrote a function `path_ends_with` that can be used to check if something ends in `Gd`. However it does not take into account generics at the moment.
   * The simple way: compare the token-stream against hardcoded `Gd<Self>` and `godot::obj::Gd<Self>`.
   * We may not support other valid variants such as `Gd<MyClass>`, `Gd<TypeAlias>` or whatever, but that should be OK.

This should come with descriptive error messages, so that a user knows what went wrong.

Reply:
How should I compare the param against that?
The type is pretty complex, I can dig into it with a match statement but then I'm not sure how to compare `Vec<TokenTree>` to the string "Gd<Self>".

Reply:
It might be a bit tedious, because `TokenStream` doesn't implement `Eq`.
But this might work:
* `TokenStream` [implements `IntoIterator`](https://docs.rs/proc-macro2/latest/proc_macro2/struct.TokenStream.html#impl-IntoIterator-for-TokenStream)
* so you can use `.collect::<Vec<TokenTree>>()` on it
* then you can compare your `venial::TyExpr` against that

So basically you could use 2 hardcoded tokenstreams `quote! { Gd<Self> }` and `quote! { godot::obj::Gd<Self> }`, which you can convert into vectors of token-trees.

If that turns out to be too inflexible or complicated, you can also try the other route I suggested:


> Alternatively, this is all delegated to the type system and instead of removing the parameter, you reuse it to declare an intermediate function. That might be more flexible, but I'm not sure if it has a positive or negative effect on error messages.

Reply:
Adding an intermediate function doesn't really seem to change the error message. In both cases it says:
```
error[E0308]: mismatched types
  --> itest\rust\src\register_tests\func_test.rs:47:1
   |
47 | #[godot_api]
   | ^^^^^^^^^^^^
   | |
   | expected `Gd<Node3D>`, found `Gd<GdSelfReference>`
   | arguments to this function are incorrect
   |
   = note: expected struct `godot::prelude::Gd<godot::prelude::Node3D>`
              found struct `godot::prelude::Gd<GdSelfReference>`
note: associated function defined here
  --> itest\rust\src\register_tests\func_test.rs:72:8
   |
72 |     fn takes_gd_as_self_error(this: Gd<Node3D>, arg1: i32, arg2: i32) -> i32 {
   |        ^^^^^^^^^^^^^^^^^^^^^^ ----------------
   = note: this error originates in the attribute macro `godot_api` (in Nightly builds, run with -Z macro-backtrace for more info)
   ```
   
 Which I think gets the point across, "we're passing in a Gd<GdSelfReference>, you took a Gd<Node3D>".
 With a better error for passing in no argument at all, I think this is fine?

Reply:
Yes, I agree, thanks a lot for checking the errors! 🙂

Reply:
Also: checking for Gd<Self> by text doesn't work because sometimes I see:
```
error[E0401]: can't use generic parameters from outer item
  --> itest\rust\src\register_tests\func_test.rs:69:20
   |
52 | #[godot_api]
   | ------------ `Self` type implicitly declared here, by this `impl`
...
69 |     fn b2(this: Gd<Self>, new_internal: i32) -> i32 {
   |                    ^^^^
   |                    |
   |                    use of generic parameter from outer item
   |                    refer to the type directly here instead
```
Sometimes it is okay to use `Gd<Self>` but sometimes you need `Gd<MyType>` explicitly. I'm not sure what exactly makes the difference though.

Reply:
Let's then leave it to the type system, I made some good experiences with that approach in [#415](https://www.github.com/godot-rust/gdext/pull/415) 😉